HPCC-33280: Fix Coverity-reported defect in CEspHttpServer

timothyklemm opened the pull request:
Address a reported concern that CEspHttpServer::m_apport must be checked for
NULL before use because a recent change included a check.
- Remove all NULL checks for m_apport, as they are currently unnecessary.
- Change required pointer parameters of the server constructor to references,
  ensuring that no NULL checks are required in the future.
- Change required pointer parameters of the CHttpThread constructor to
  references, ensuring the server's caller does not require NULL checks.
- Refactor the initialization of CPooledHttpThread instances to improve type
  safety of all parameter values, including the use of references for required
  instances.
- Ensure that pooled thread exceptions are handled consistently by both
  CHttpProtocol and CSecureHttpProtocol with respect to socket cleanup.

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

Commit:
HPCC-33280: Fix Coverity-reported defect in CEspHttpServer

Address a reported concern that CEspHttpServer::m_apport must be checked for
NULL before use because a recent change included a check.
- Remove all NULL checks for m_apport, as they are currently unnecessary.
- Change required pointer parameters of the server constructor to references,
  ensuring that no NULL checks are required in the future.
- Change required pointer parameters of the CHttpThread constructor to
  references, ensuring the server's caller does not require NULL checks.
- Refactor the initialization of CPooledHttpThread instances to improve type
  safety of all parameter values, including the use of references for required
  instances.
- Ensure that pooled thread exceptions are handled consistently by both
  CHttpProtocol and CSecureHttpProtocol with respect to socket cleanup.

Signed-off-by: Tim Klemm <Tim.Klemm@lexisnexisrisk.com>

rpastrana reviewed (commented):
@timothyklemm the changes seem* fine. However it's not abundantly clear what benefit this change provides. It's not clear from the changes how m_apport is safe to use in all code paths.

timothyklemm commented:
The protocol classes throw exceptions if the pointer is NULL before creating threads, which in turn create servers. The thread and server classes already make multiple assumptions about the pointer not being NULL. The server method from which I removed the two checks for NULL starts by dereferencing the pointer without checking.

The issue came up because my change to the server span creation time inserted an unnecessary check which was flagged by the most recent Coverity scan. If it was necessary to check where I added the check, then it would also be necessary to check before subsequent references. Unfortunately, the scan hasn't been able to point out that if the pre-existing check was necessary, all of the preceding references would also require checks.

rpastrana reviewed (commented):
@timothyklemm left a few questions/concerns.
Overall this seems like a good change, but it doesn't seem to match the commit title. Let's make sure the commit title and message match the changes and informs the reviewer.

timothyklemm commented on esp/bindings/http/platform/httpprot.cpp:
```diff
@@ -137,42 +162,17 @@ bool CHttpProtocol::notifySelected(ISocket *sock,unsigned selected, IPersistentH
+                    pti.persistentHandler = persistentHandler;
+                    pti.shouldClose = shouldClose;
+                    // cleanup on exception is handled by pti
+                    http_thread_pool->start((void*)&pti, "", m_threadCreateTimeout > 0?m_threadCreateTimeout*1000:0);
```
Based on the original code, the answer is no. Instead of catching, reacting to, and re-throwing all exceptions, pti's destructor will react to the existence of an exception without capturing it.

As for the destructor's abbreviated handler relative to what was here, there is no longer a socket reference to be released nor is there a heap allocation to be deleted.

timothyklemm commented on esp/bindings/http/platform/httpprot.cpp:
```diff
@@ -41,6 +41,31 @@
+    ~PooledThreadInfo()
+    {
+#if __cplusplus >= 201703L
+        if (std::uncaught_exceptions() > 0)
```
My understanding is that it is per thread. Exceptions could be captured if the destructor needed to know details about what caused the stack to unwind. In this instance, we were not showing interest in what caused the failure.

timothyklemm replied:
As implemented, a double close is safe - the first close clears a value that must be set for the second close to have an effect. 

We already have an example of not explicitly closing the socket, because the exception isn't caught early enough. It has been suggested to me that preceding the close with an explicit shutdown could be an improvement. Reinforces my opinion that never closing is incorrect behavior.

Could somebody catch the exception because the comment stating that exception cleanup was happening elsewhere was not clear enough? Yes. Is it likely enough to warrant pre-empting it? I'll defer to the reviewers.

timothyklemm commented on esp/bindings/http/platform/httpprot.cpp:
```diff
@@ -41,6 +41,31 @@
+    PooledThreadInfo(ISocket& _socket, CEspApplicationPort& _apport) : socket(_socket), apport(_apport) {}
+    ~PooledThreadInfo()
+    {
+#if __cplusplus >= 201703L
```
We had no error handling in the secure protocol. The socket was not being closed and now will be. The structure and its usage address two unhandled memory leaks.

I wrote a simple test on my Mac, because a vcpkg patch file states that this function can't be used in Apple builds (and I hadn't noticed it was already being used in the platform). A destructor inside a try block observed the exception and the catch block still caught it. For this case, we're observing without interfering with the standard stack unwind behavior.

timothyklemm commented on esp/bindings/http/platform/httpprot.cpp:
```diff
@@ -137,42 +162,17 @@ bool CHttpProtocol::notifySelected(ISocket *sock,unsigned selected, IPersistentH
-                        throw;
-                    }
-                    delete [] holder;
+                    PooledThreadInfo pti(*accepted, *apport);
```
Agreed.

timothyklemm commented on esp/bindings/http/platform/httpservice.cpp:
```diff
@@ -525,134 +523,131 @@ int CEspHttpServer::processRequest()
+            else
+                unsupported(method);
         }
+        ctx->addTraceSummaryTimeStamp(LogMin, "handleHttp");
```
Correct,

timothyklemm commented on esp/bindings/http/platform/httpservice.cpp:
```diff
@@ -430,8 +430,6 @@ int CEspHttpServer::processRequest()
-        else if (!m_apport)
-            wantTracing = false;
         Owned<ISpan> serverSpan;
         if (wantTracing)
```
The flag is affected by tracing being enabled and also by certain "esp" service GET requests that were processed prior to the original creation of the span (look just before the start of processRequest to see the method names that process without tracing).

asselitx commented on esp/bindings/http/platform/httpprot.cpp:
```diff
@@ -515,13 +495,16 @@ void CHttpThread::returnSocket()
+        throw makeStringException(-1, "CPooledHttpThread::init: invalid param");
+    PooledThreadInfo* ptip = (PooledThreadInfo*)param;
+    m_socket.set(&ptip->socket);
+    m_apport = &ptip->apport;
```
I'm guessing you left `m_apport` as a pointer to avoid touching a lot more code? Or is there positive benefit for checking an uninitialized thread? It seems like it would be clearer to change it to a reference, but I understand that is probably very low priority and not something to tackle in this PR.

asselitx reviewed (approved):
This looks good as long as none of my questions about the exception throwing/catching  warrant a change.

Like Rodrigo said, it would be good to test thoroughly locally including leaks and exception conditions. Check with Mark and Attila to see if any existing tests cover this code, and if not, add some if it is reasonable (esp. throwing an exception that would cause socket close).

ghalliday commented on esp/bindings/http/platform/httpprot.cpp:
```diff
@@ -41,6 +41,31 @@
+    PooledThreadInfo(ISocket& _socket, CEspApplicationPort& _apport) : socket(_socket), apport(_apport) {}
+    ~PooledThreadInfo()
+    {
+#if __cplusplus >= 201703L
```
We are always compiling with c++17, so I don't think the #if is necessary.  Is it?

timothyklemm replied:
Probably not. I found the inverse condition in hqlhtcpp.cpp tried to stay consistent. Will remove.

ghalliday commented on esp/bindings/http/platform/httpprot.cpp:
```diff
@@ -533,6 +516,11 @@ CPooledHttpThread::~CPooledHttpThread()
 {
+    if (!m_socket || !m_apport)
+    {
+        IERRLOG("CPooledHttpThread::thradmain: uninitialized thread");
```
typo: threadmain

ghalliday reviewed (commented):
@timothyklemm I started off thinking this was a bit unnecessary given the jira, then I saw the original code!  A big improvement.
A couple of minor comments.

timothyklemm commented:
valgrind has reported multiple instances of what I can only describe as false positives. A combination of log analysis and debugging prove that reported leaks are not leaks.
- The `holder` variable was allegedly leaked on exception, but log analysis proved the line following the deletion executed.
- Unpooled threads are allegedly leaked after successful requests, both before and after my change, but debugging showed the thread being unconditionally released and deleted.
- Pooled threads are allegedly retaining their socket references after successful requests. I believe this to be inaccurate, but it could happen in some failure scenarios. Will make a change that satisfies valgrind in the success case and also handles the failure cases.

rpastrana commented on esp/bindings/http/platform/httpprot.cpp:
```diff
@@ -516,9 +512,18 @@ CPooledHttpThread::~CPooledHttpThread()
+        CPooledHttpThread& owner;
+        SocketReturner(CPooledHttpThread& _owner) : owner(_owner) {}
+        ~SocketReturner() { owner.returnSocket();}
+    } sr(*this);
```
so what's the magic here? what is struct sr doing?

timothyklemm replied:
It's forcing the socket to be returned for every exit path of the method. I had tried creating a new method to do the work of threadmain, leaving thread main to do nothing but call that method and return the socket. Valgrind still complained about leaked sockets. This was the approach that made it happy.

ghalliday reviewed (approved):
Please squash

timothyklemm commented:
@ghalliday squashed...
